Allow overriding unstruct_collection_overrides in preconf make_converter

* cattrs version: 22.2.0
* Python version: 3.9.16
* Operating System: macOS

### Description

What I was trying to do:
```python
from collections.abc import Set

from cattrs.preconf.json import make_converter as json_make_converter

json_converter = json_make_converter(unstruct_collection_overrides={Set: sorted})
```

This didn't work because `cattrs.preconf.json.make_converter` overrides the `kwargs`: https://github.com/python-attrs/cattrs/blob/5ccd616cb9ae2ad88b5d4ba7816f7fad3a9f84be/src/cattrs/preconf/json.py#L40-L44

Then I tried:
```python
from collections.abc import Set

from cattrs.preconf.json import make_converter as json_make_converter

json_converter = json_make_converter().copy(unstruct_collection_overrides={Set: sorted})
```

But later in the code I call `json_converter.dumps` and mypy complained `error: "Converter" has no attribute "dumps"  [attr-defined]` because `copy` returns a `Converter`.

### What I Did
```python
from collections.abc import Set

from cattrs.preconf.json import JsonConverter
from cattrs.preconf.json import configure_converter as json_configure_converter

json_converter = JsonConverter(unstruct_collection_overrides={Set: sorted})
json_configure_converter(json_converter)
```

Maybe the two forms that I've tried above should also be allowed?

--- a/HISTORY.md
+++ b/HISTORY.md
@@ -14,6 +14,8 @@
   ([#81](https://github.com/python-attrs/cattrs/issues/81))
 - Add `cbor2` serialization library to the `cattr.preconf` package.
 - Add optional dependencies for `cattrs.preconf` third-party libraries. ([#337](https://github.com/python-attrs/cattrs/pull/337))
+- All preconf converters now allow overriding the default `unstruct_collection_overrides` in `make_converter`.
+  ([#350](https://github.com/python-attrs/cattrs/issues/350) [#353](https://github.com/python-attrs/cattrs/pull/353))
 
 ## 22.2.0 (2022-10-03)
 

--- a/src/cattrs/preconf/bson.py
+++ b/src/cattrs/preconf/bson.py
@@ -88,8 +88,8 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> BsonConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = BsonConverter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/cbor2.py
+++ b/src/cattrs/preconf/cbor2.py
@@ -34,8 +34,8 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> Cbor2Converter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = Cbor2Converter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/json.py
+++ b/src/cattrs/preconf/json.py
@@ -38,9 +38,9 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> JsonConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
         Counter: dict,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = JsonConverter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/msgpack.py
+++ b/src/cattrs/preconf/msgpack.py
@@ -34,8 +34,8 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> MsgpackConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = MsgpackConverter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/orjson.py
+++ b/src/cattrs/preconf/orjson.py
@@ -68,8 +68,8 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> OrjsonConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = OrjsonConverter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/pyyaml.py
+++ b/src/cattrs/preconf/pyyaml.py
@@ -35,8 +35,8 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> PyyamlConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         FrozenSetSubscriptable: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = PyyamlConverter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/tomlkit.py
+++ b/src/cattrs/preconf/tomlkit.py
@@ -67,9 +67,9 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> TomlkitConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
         tuple: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = TomlkitConverter(*args, **kwargs)
     configure_converter(res)

--- a/src/cattrs/preconf/ujson.py
+++ b/src/cattrs/preconf/ujson.py
@@ -39,8 +39,8 @@ def configure_converter(converter: BaseConverter):
 
 def make_converter(*args, **kwargs) -> UjsonConverter:
     kwargs["unstruct_collection_overrides"] = {
-        **kwargs.get("unstruct_collection_overrides", {}),
         AbstractSet: list,
+        **kwargs.get("unstruct_collection_overrides", {}),
     }
     res = UjsonConverter(*args, **kwargs)
     configure_converter(res)

--- a/tests/test_preconf.py
+++ b/tests/test_preconf.py
@@ -26,8 +26,10 @@ from hypothesis.strategies import (
 )
 
 from cattrs._compat import (
+    AbstractSet,
     Counter,
     FrozenSet,
+    FrozenSetSubscriptable,
     Mapping,
     MutableMapping,
     MutableSequence,
@@ -173,6 +175,15 @@ def test_stdlib_json_converter(everything: Everything):
     assert converter.loads(converter.dumps(everything), Everything) == everything
 
 
+@given(everythings())
+def test_stdlib_json_converter_unstruct_collection_overrides(everything: Everything):
+    converter = json_make_converter(unstruct_collection_overrides={AbstractSet: sorted})
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
+
+
 @given(
     everythings(
         min_int=-9223372036854775808, max_int=9223372036854775807, allow_inf=False
@@ -196,6 +207,21 @@ def test_ujson_converter(everything: Everything):
     converter = ujson_make_converter()
     raw = converter.dumps(everything)
     assert converter.loads(raw, Everything) == everything
+
+
+@given(
+    everythings(
+        min_int=-9223372036854775808, max_int=9223372036854775807, allow_inf=False
+    )
+)
+def test_ujson_converter_unstruct_collection_overrides(everything: Everything):
+    converter = ujson_make_converter(
+        unstruct_collection_overrides={AbstractSet: sorted}
+    )
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
 
 
 @pytest.mark.skipif(python_implementation() == "PyPy", reason="no orjson on PyPy")
@@ -231,6 +257,24 @@ def test_orjson_converter(everything: Everything, detailed_validation: bool):
     assert converter.loads(raw, Everything) == everything
 
 
+@pytest.mark.skipif(python_implementation() == "PyPy", reason="no orjson on PyPy")
+@given(
+    everythings(
+        min_int=-9223372036854775808, max_int=9223372036854775807, allow_inf=False
+    )
+)
+def test_orjson_converter_unstruct_collection_overrides(everything: Everything):
+    from cattrs.preconf.orjson import make_converter as orjson_make_converter
+
+    converter = orjson_make_converter(
+        unstruct_collection_overrides={AbstractSet: sorted}
+    )
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
+
+
 @given(everythings(min_int=-9223372036854775808, max_int=18446744073709551615))
 def test_msgpack(everything: Everything):
     from msgpack import dumps as msgpack_dumps
@@ -249,6 +293,17 @@ def test_msgpack_converter(everything: Everything):
     converter = msgpack_make_converter()
     raw = converter.dumps(everything)
     assert converter.loads(raw, Everything, strict_map_key=False) == everything
+
+
+@given(everythings(min_int=-9223372036854775808, max_int=18446744073709551615))
+def test_msgpack_converter_unstruct_collection_overrides(everything: Everything):
+    converter = msgpack_make_converter(
+        unstruct_collection_overrides={AbstractSet: sorted}
+    )
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
 
 
 @given(
@@ -294,6 +349,22 @@ def test_bson_converter(everything: Everything, detailed_validation: bool):
     )
 
 
+@given(
+    everythings(
+        min_int=-9223372036854775808,
+        max_int=9223372036854775807,
+        allow_null_bytes_in_keys=False,
+        allow_datetime_microseconds=False,
+    )
+)
+def test_bson_converter_unstruct_collection_overrides(everything: Everything):
+    converter = bson_make_converter(unstruct_collection_overrides={AbstractSet: sorted})
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
+
+
 @given(everythings())
 def test_pyyaml(everything: Everything):
     from yaml import safe_dump, safe_load
@@ -309,6 +380,15 @@ def test_pyyaml_converter(everything: Everything):
     converter = pyyaml_make_converter()
     raw = converter.dumps(everything)
     assert converter.loads(raw, Everything) == everything
+
+
+@given(everythings())
+def test_pyyaml_converter_unstruct_collection_overrides(everything: Everything):
+    converter = pyyaml_make_converter(
+        unstruct_collection_overrides={FrozenSetSubscriptable: sorted}
+    )
+    raw = converter.unstructure(everything)
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
 
 
 @given(
@@ -345,6 +425,24 @@ def test_tomlkit_converter(everything: Everything, detailed_validation: bool):
     assert converter.loads(raw, Everything) == everything
 
 
+@given(
+    everythings(
+        min_key_length=1,
+        allow_null_bytes_in_keys=False,
+        key_blacklist_characters=['"', "\\"],
+        allow_control_characters_in_values=False,
+    )
+)
+def test_tomlkit_converter_unstruct_collection_overrides(everything: Everything):
+    converter = tomlkit_make_converter(
+        unstruct_collection_overrides={AbstractSet: sorted}
+    )
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])
+
+
 def test_bson_objectid():
     """BSON ObjectIds are supported by default."""
     converter = bson_make_converter()
@@ -368,3 +466,14 @@ def test_cbor2_converter(everything: Everything):
     converter = cbor2_make_converter()
     raw = converter.dumps(everything)
     assert converter.loads(raw, Everything) == everything
+
+
+@given(everythings(min_int=-9223372036854775808, max_int=18446744073709551615))
+def test_cbor2_converter_unstruct_collection_overrides(everything: Everything):
+    converter = cbor2_make_converter(
+        unstruct_collection_overrides={AbstractSet: sorted}
+    )
+    raw = converter.unstructure(everything)
+    assert raw["a_set"] == sorted(raw["a_set"])
+    assert raw["a_mutable_set"] == sorted(raw["a_mutable_set"])
+    assert raw["a_frozenset"] == sorted(raw["a_frozenset"])

